File: src/actions_cpp/src/count_until_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "my_robot_interfaces/action/count_until.hpp"

class CountUntilServerNode : public rclcpp::Node
{
public:
    using CountUntil = my_robot_interfaces::action::CountUntil;
    using GoalHandleCountUntil = rclcpp_action::ServerGoalHandle<CountUntil>;

    CountUntilServerNode() : Node("count_until_server")
    {
        this->goal_handle_ = nullptr;
        this->callback_group_ = this->create_callback_group(rclcpp::CallbackGroupType::Reentrant);

        this->action_server_ = rclcpp_action::create_server<CountUntil>(
            this,
            "count_until",
            std::bind(&CountUntilServerNode::handle_goal, this, std::placeholders::_1, std::placeholders::_2),
            std::bind(&CountUntilServerNode::handle_cancel, this, std::placeholders::_1),
            std::bind(&CountUntilServerNode::handle_accepted, this, std::placeholders::_1),
            rcl_action_server_get_default_options(), 
            callback_group_);
        
        RCLCPP_INFO(this->get_logger(), "Count until server has been started");
    }

private:
    rclcpp_action::Server<CountUntil>::SharedPtr action_server_;
    rclcpp::CallbackGroup::SharedPtr callback_group_;
    std::shared_ptr<GoalHandleCountUntil> goal_handle_;
    std::mutex mutex_;
    rclcpp_action::GoalUUID preempting_goal_uuid_;

    rclcpp_action::GoalResponse handle_goal(const rclcpp_action::GoalUUID &uuid, std::shared_ptr<const CountUntil::Goal> goal)
    {
        (void)uuid;

        if (goal->target_number < 0 || goal->period < 0.0)
        {
            RCLCPP_WARN(this->get_logger(), "Rejected goal request with negative target_number %ld or period %.2f", goal->target_number, goal->period);
            return rclcpp_action::GoalResponse::REJECT;
        }

        // Policy: Refuse new goal if current goal is active
        // {
        //     std::lock_guard<std::mutex> lock(mutex_);
        //     if (goal_handle_ && goal_handle_->is_active())
        //     {
        //         RCLCPP_WARN(this->get_logger(), "Rejected new goal request, current goal is active");
        //         return rclcpp_action::GoalResponse::REJECT;
        //     }
        // }

        // Policy: preempt the currently active goal when receiving a new valid one
        {
            std::lock_guard<std::mutex> lock(mutex_);   
            if (goal_handle_ && goal_handle_->is_active())
            {
                preempting_goal_uuid_ = goal_handle_->get_goal_id();
            }
        }

        RCLCPP_INFO(this->get_logger(), "Received goal request with target_number %ld, period %.2f", goal->target_number, goal->period);
        return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
    }

    rclcpp_action::CancelResponse handle_cancel(const std::shared_ptr<GoalHandleCountUntil> goal_handle)
    {
        RCLCPP_INFO(this->get_logger(), "Received request to cancel goal");
        (void)goal_handle;
        return rclcpp_action::CancelResponse::ACCEPT;
    }

    void handle_accepted(const std::shared_ptr<GoalHandleCountUntil> goal_handle)
    {
        using namespace std::placeholders;
        execute_goal(goal_handle);
        // std::thread{std::bind(&CountUntilServerNode::execute, this, _1), goal_handle}.detach();
    }

    void execute_goal(const std::shared_ptr<GoalHandleCountUntil> goal_handle)
    {
        {
            std::lock_guard<std::mutex> lock(mutex_);
            goal_handle_ = goal_handle;
        }

        RCLCPP_INFO(this->get_logger(), "Executing goal for target_number %ld, period %.2f", goal_handle->get_goal()->target_number, goal_handle->get_goal()->period);
        const auto goal = goal_handle->get_goal();
        auto feedback = std::make_shared<CountUntil::Feedback>();
        auto &progress = feedback->current_number;
        auto result = std::make_shared<CountUntil::Result>();

        rclcpp::Rate loop_rate(1.0 / goal->period);

        for (int i = 1; i <= goal->target_number; ++i)
        {
            if(preempting_goal_uuid_ == goal_handle->get_goal_id())
            {
                result->reached_number = progress;
                goal_handle->abort(result);
                RCLCPP_INFO(this->get_logger(), "Preempted goal for target_number %ld, period %.2f", goal->target_number, goal->period);
                return;
            }
            if (goal_handle->is_canceling())
            {
                result->reached_number = progress;
                goal_handle->canceled(result);
                RCLCPP_INFO(this->get_logger(), "Goal canceled for target_number %ld, period %.2f", goal->target_number, goal->period);
                return;
            }

            progress = i;
            goal_handle->publish_feedback(feedback);
            RCLCPP_INFO(this->get_logger(), "Published feedback: %ld", progress);
            loop_rate.sleep();
        }

        if (rclcpp::ok())
        {
            result->reached_number = progress;
            goal_handle->succeed(result);
            RCLCPP_INFO(this->get_logger(), "Goal succeeded for target_number %ld, period %.2f", goal->target_number, goal->period);
        }
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<CountUntilServerNode>();
    rclcpp::executors::MultiThreadedExecutor executor;
    executor.add_node(node);
    executor.spin();
    rclcpp::shutdown();
    return 0;
}

File: src/my_py_pkg/my_py_pkg/count_until_server.py
import rclpy
import time
import threading
from rclpy.node import Node
from rclpy.action import ActionServer
from my_robot_interfaces.action import CountUntil
from rclpy.action import CancelResponse, GoalResponse
from rclpy.action.server import ServerGoalHandle
from rclpy.executors import MultiThreadedExecutor
from rclpy.callback_groups import ReentrantCallbackGroup

class CountUntilServerNode(Node):

    def __init__(self):
        super().__init__('count_until_server')
        self._action_server = ActionServer(
            self,
            CountUntil,
            'count_until',
            execute_callback=self.execute_callback,
            handle_accepted_callback=self.handle_accepted_callback,
            goal_callback=self.goal_callback,
            cancel_callback=self.cancel_callback,
            callback_group=ReentrantCallbackGroup())
        
        self._goal_lock = threading.Lock()
        self._goal_handle = None

        self._goal_queue = []

        self.get_logger().info('Action server ros2 node for count until has been started')

    def goal_callback(self, goal_request : CountUntil.Goal):
        self.get_logger().info('Received goal request with target number {0} and period {1}'.format(goal_request.target_number, goal_request.period))
        
        # Policy: refuse new goal if current goal still active
        # with self._goal_lock:
        #     if self._goal_handle is not None and self._goal_handle.is_active:
        #         self.get_logger().info('A goal is already active: goal request rejected')
        #         return GoalResponse.REJECT
        
        if goal_request.period < 0 or goal_request.target_number < 0:
            self.get_logger().info('Goal request rejected')
            return GoalResponse.REJECT

        # Policy: preempt currently active goal if any
        # with self._goal_lock:
        #     if self._goal_handle is not None and self._goal_handle.is_active:
        #         self.get_logger().info('Preempting the current goal active and accept new goal')
        #         self._goal_handle.abort()
        
        return GoalResponse.ACCEPT

    def handle_accepted_callback(self, goal_handle : ServerGoalHandle):
        self.get_logger().info('Goal has been accepted')
        with self._goal_lock:
            if self._goal_handle is not None:
                self._goal_queue.append(goal_handle)
                self.get_logger().info('Pushing to the queue accepted goal {0}, new queue size = {1}'.format(goal_handle.request, len(self._goal_queue)))
            else:
                self._goal_handle = goal_handle
                goal_handle.execute()

    def cancel_callback(self, goal_handle : ServerGoalHandle):
        self.get_logger().info('Received cancel request')
        return CancelResponse.ACCEPT

    def execute_callback(self, goal_handle : ServerGoalHandle):
        self.get_logger().info('Executing goal {0}...'.format(goal_handle.request))
        
        # Get the goal request
        target_number = goal_handle.request.target_number
        period = goal_handle.request.period
        counter = 0
        
        feedback = CountUntil.Feedback()
        result = CountUntil.Result()
        # Execute the action
        for i in range(target_number):
            if not goal_handle.is_active:
                self.get_logger().info('Goal has been preempted')
                result.reached_number = counter
                self.process_next_goal_in_queue()
                return result
            
            elif goal_handle.is_cancel_requested:
                self.get_logger().info('Goal has been canceled')
                result.reached_number = counter
                goal_handle.canceled()
                self.process_next_goal_in_queue()
                return result
            
            counter += 1
            self.get_logger().info('Count: {0}'.format(counter))
            feedback.current_number = counter
            goal_handle.publish_feedback(feedback)
            time.sleep(period)

        # Once done, set goal final state
        goal_handle.succeed()

        # and send the result
        result.reached_number = counter
        self.process_next_goal_in_queue()
        return result

    def process_next_goal_in_queue(self):
        with self._goal_lock:
            if len(self._goal_queue) > 0:
                self._goal_handle = self._goal_queue.pop(0)
                self.get_logger().info('Processing next goal in the queue {0}, new queue size = {1}'.format(self._goal_handle.request, len(self._goal_queue)))
                self._goal_handle.execute()
            else:
                self._goal_handle = None

def main(args=None):
    rclpy.init(args=args)
    action_server = CountUntilServerNode()
    rclpy.spin(action_server, MultiThreadedExecutor())
    rclpy.shutdown()

if __name__ == '__main__':
    main()